### PR TITLE
feat(Emoji): create `getURL()` function to get more extension of emoji

### PR DESCRIPTION
**Pull request by dager-mohamed**
**Please describe the changes this PR makes and why it should be merged:**

discordjs support only `.gif` and `.png` for emoji extension, so I created the `getURL(ext)` function to make the user get any extension of emoji he wants.

**Status and versioning classification:**

Please move lines that apply to you out of the comment:
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

File: packages/discord.js/src/structures/Emoji.js [hunk 1]
```diff
@@ -47,13 +47,18 @@ class Emoji extends Base {
...
+   * @returns {string}
    */
-  get url() {
-    return this.id && this.client.rest.cdn.emoji(this.id, this.animated ? 'gif' : 'png');
+  url(extension) {
+    let final_result;
+    const extensions = ['webp', 'png', 'jpg', 'jpeg', 'gif'];
+    if (!extensions.includes(extension)) extension = this.animated ? 'gif' : 'png';
+    final_result = this.id && this.client.rest.cdn.emoji(this.id, extension);
+    return final_result;
```
Comment (thehackerboi69github) [inline packages/discord.js/src/structures/Emoji.js:61]:
```suggestion
    const extensions = ['webp', 'png', 'jpg', 'jpeg', 'gif'];
    if (!extensions.includes(extension)) extension = this.animated ? 'gif' : 'png';
    return this.id && this.client.rest.cdn.emoji(this.id, extension);
```
You're only assigning to `final_result` once.

**Review by thehackerboi69github — CHANGES_REQUESTED** [review]
I'm not so sure about this one. This breaks consistency with other `url` getters in the library, and if users want to have another extension, they can use `client.rest.cdn.emoji(emoji.id, 'jpg')` for example. This function is a shortcut for that, but it breaks consistency, so I don't think it should be added like this.

File: packages/discord.js/src/structures/Emoji.js (hunk 1 above)
Comment (dager-mohamed) [reply packages/discord.js/src/structures/Emoji.js:61]: Good job bro

**Comment by jaw0r3k** [comment]
I agree with thehackerboi that .url should be a getter
So making function like getUrl would be better

**Comment by dager-mohamed** [comment]
> I agree with thehackerboi that .url should be a getter So making a function like getUrl would be better

so I have an idea and I don't know if this idea is like you said or not,

I can create `getUrl(URL)` as you said function inside the class to check the URL of the emoji and then return the extension, 
and I will use the `getUrl()` function inside `url()` and it will be getter that you mean?

I hope that you mean.
سخ 
سخ

**Comment by jaw0r3k** [comment]
Nah i just mean getURL(ext) which will return a url
and leave url as getter ( like everywhere else )

**Comment by dager-mohamed** [comment]
> 

ahaa ok i understand now, i will do it

**Comment by dager-mohamed** [comment]
now i added the `getUrl(ext)` function and `url` is a getter now.

File: packages/discord.js/src/structures/Emoji.js [hunk 2]
```diff
@@ -56,6 +56,17 @@ class Emoji extends Base {
...
   }
 
+  /**
+   * @type {?string}
+   * @param {ImageExtension} ext the extension of emoji
+   * @returns {string}
+   */
+  getUrl(ext) {
+    const extensions = ['webp', 'png', 'jpg', 'jpeg', 'gif'];
+    if (!extensions.includes(ext) || !ext) return undefined;
```
Comment (Syjalo) [inline packages/discord.js/src/structures/Emoji.js:66]: Do we need that? The rest cdn getter throws an error if the extension is not valid.

**Review by Syjalo — CHANGES_REQUESTED** [review]

File: packages/discord.js/src/structures/Emoji.js (hunk 2 above)
Comment (dager-mohamed) [reply packages/discord.js/src/structures/Emoji.js:66]: > Do we need that? The rest cdn getter throws an error if the extension is not valid.

are you talking about the all `getUrl` function or you talking about this
```js
    const extensions = ['webp', 'png', 'jpg', 'jpeg', 'gif'];
    if (!extensions.includes(ext) || !ext) return undefined;
    ````
    
    anyway I made this to return undefined like `url` getter , anyway if it is not required we can remove it

File: packages/discord.js/src/structures/Emoji.js [hunk 3]
```diff
@@ -56,6 +56,17 @@ class Emoji extends Base {
...
+    return this.id && this.client.rest.cdn.emoji(this.id, ext);
```
Comment (thehackerboi69github) [inline packages/discord.js/src/structures/Emoji.js:67]: I think it's best to throw an error when the user provides an invalid extension.

File: packages/discord.js/src/structures/Emoji.js [hunk 4]
```diff
@@ -56,6 +56,17 @@ class Emoji extends Base {
     return this.id && this.client.rest.cdn.emoji(this.id, this.animated ? 'gif' : 'png');
   }
 
+  /**
+   * @type {?string}
+   * @param {ImageExtension} ext the extension of emoji
+   * @returns {string}
+   */
+  getUrl(ext) {
```
Comment (thehackerboi69github) [inline packages/discord.js/src/structures/Emoji.js:64]:
```suggestion
  getURL(ext) {
```

File: packages/discord.js/src/structures/Emoji.js (hunk 3 above)
Comment (dager-mohamed) [reply packages/discord.js/src/structures/Emoji.js:67]: you are right i will do it

**Comment by dager-mohamed** [comment]
note: the rest.cdn is automatically throw the error at all

**Comment by thehackerboi69github** [comment]
> note: the rest.cdn is automatically throw the error at all

In that case you don't have to return undefined if an invalid extension is provided.

**Review by thehackerboi69github — APPROVED** [review]

File: packages/discord.js/src/structures/Emoji.js [hunk 5]
```diff
@@ -56,6 +56,15 @@ class Emoji extends Base {
...
   }
 
+  /**
+   * @type {?string}
+   * @param {ImageExtension} ext the extension of emoji
+   * @returns {string}
+   */
+  getURL(ext) {
+    return this.id && this.client.rest.cdn.emoji(this.id, ext);
+  }
```
Comment (NotSugden) [inline packages/discord.js/src/structures/Emoji.js:66]: i think the position of this function needs to be below all of the getter methods
```suggestion
  /**
   * @param {ImageExtension} ext the extension of emoji
   * @returns {?string}
   */
  getURL(ext) {
    return this.id && this.client.rest.cdn.emoji(this.id, ext);
  }
```
also corrected documentation

File: packages/discord.js/typings/index.d.ts [hunk 6]
```diff
@@ -1070,6 +1070,7 @@ export class Emoji extends Base {
   public name: string | null;
   public get identifier(): string;
   public get url(): string | null;
+  public getURL(ext: ImageExtension): string;
```
Comment (NotSugden) [inline packages/discord.js/typings/index.d.ts:1073]:
```suggestion
  public getURL(ext: ImageExtension): string | null;
```
as per previous suggestion the method can return null

**Review by NotSugden — CHANGES_REQUESTED** [review]

File: packages/discord.js/src/structures/Emoji.js (hunk 5 above)
Comment (jaw0r3k) [reply packages/discord.js/src/structures/Emoji.js:66]:
```suggestion
  /**
   * @param {ImageExtension} ext The extension of the emoji
   * @returns {?string}
   */
  getURL(ext) {
    return this.id && this.client.rest.cdn.emoji(this.id, ext);
  }
```

**Comment by Jiralite** [comment]
Superseded by #9788.

**Pull request closed without merging**